File: ios/NorthAx/Models/MetricSourcePriority.swift
import Foundation

/// A data source that can report wellness metrics. `intervals` already
/// aggregates Garmin/Strava/Apple Health upstream; `healthkit` is read on-device;
/// `manual` is reserved for user-entered values (a backend zone, not yet merged
/// client-side — see docs/multi-source-metrics.md).
enum MetricSource: String, Codable, CaseIterable, Identifiable {
    case intervals, healthkit, manual

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .intervals: return "intervals.icu"
        case .healthkit: return "Apple Health"
        case .manual:    return "Manual entry"
        }
    }
}

/// Metrics that more than one source can report, so they need conflict
/// resolution. Training load (CTL/ATL) is intervals-only and intentionally absent.
enum MergeableMetric: String, Codable, CaseIterable, Identifiable {
    case hrv, restingHR, sleep, bodyWeight

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .hrv:        return "Heart Rate Variability"
        case .restingHR:  return "Resting Heart Rate"
        case .sleep:      return "Sleep"
        case .bodyWeight: return "Body Weight"
        }
    }

    /// Sources that can actually produce this metric. The UI only ranks these,
    /// and the resolver never considers a source outside this list.
    var candidateSources: [MetricSource] {
        switch self {
        case .hrv, .restingHR, .sleep: return [.intervals, .healthkit, .manual]
        case .bodyWeight:              return [.healthkit, .manual]  // intervals doesn't carry weight
        }
    }
}

/// Per-metric ordered source preference (highest first). The first source that
/// has a value for a given day wins.
struct MetricSourcePriority: Codable, Equatable {
    /// `MergeableMetric.rawValue` → ordered list of sources.
    var order: [String: [MetricSource]]

    /// Defaults to each metric's candidate order, i.e. intervals.icu wins —
    /// identical to the app's pre-existing behavior.
    static var `default`: MetricSourcePriority {
        var o: [String: [MetricSource]] = [:]
        for m in MergeableMetric.allCases { o[m.rawValue] = m.candidateSources }
        return MetricSourcePriority(order: o)
    }

    func sources(for metric: MergeableMetric) -> [MetricSource] {
        order[metric.rawValue] ?? metric.candidateSources
    }

    /// Promote `source` to the top of `metric`'s ranking, keeping the rest in order.
    mutating func setPrimary(_ source: MetricSource, for metric: MergeableMetric) {
        var list = sources(for: metric)
        list.removeAll { $0 == source }
        list.insert(source, at: 0)
        order[metric.rawValue] = list
    }

    /// `metric.rawValue -> [source.rawValue]`, for syncing to the backend.
    var wire: [String: [String]] { order.mapValues { $0.map(\.rawValue) } }
}

// MARK: - Wire form (backend `user_preferences.metric_priority`)
// In an extension so the memberwise `init(order:)` is preserved.
extension MetricSourcePriority {
    /// Rebuild from the wire form, filling any missing metric with its defaults
    /// so newly-added metrics are always present.
    init(wire: [String: [String]]) {
        var o: [String: [MetricSource]] = [:]
        for (metric, sources) in wire { o[metric] = sources.compactMap(MetricSource.init) }
        for m in MergeableMetric.allCases where o[m.rawValue] == nil { o[m.rawValue] = m.candidateSources }
        self.init(order: o)
    }
}

File: android/app/src/main/java/app/northax/domain/model/MetricSourcePriority.kt
package app.northax.domain.model

import kotlinx.serialization.Serializable

/**
 * A data source that can report wellness metrics. `intervals` already
 * aggregates Garmin/Strava upstream; `manual` is user-entered values.
 * (HealthKit exists in the iOS app only; Android ranks intervals vs manual.)
 */
@Serializable
enum class MetricSource(val raw: String) {
    Intervals("intervals"),
    Healthkit("healthkit"),
    Manual("manual");

    val displayName: String
        get() = when (this) {
            Intervals -> "intervals.icu"
            Healthkit -> "Apple Health"
            Manual -> "Manual entry"
        }

    companion object {
        fun fromRaw(raw: String): MetricSource? = entries.firstOrNull { it.raw == raw }
    }
}

/**
 * Metrics that more than one source can report, so they need conflict
 * resolution. Training load (CTL/ATL) is intervals-only and intentionally absent.
 */
enum class MergeableMetric(val raw: String) {
    Hrv("hrv"),
    RestingHR("restingHR"),
    Sleep("sleep"),
    BodyWeight("bodyWeight");

    val displayName: String
        get() = when (this) {
            Hrv -> "Heart Rate Variability"
            RestingHR -> "Resting Heart Rate"
            Sleep -> "Sleep"
            BodyWeight -> "Body Weight"
        }

    /** Sources that can actually produce this metric on Android. */
    val candidateSources: List<MetricSource>
        get() = when (this) {
            Hrv, RestingHR, Sleep -> listOf(MetricSource.Intervals, MetricSource.Manual)
            BodyWeight -> listOf(MetricSource.Manual) // intervals doesn't carry weight
        }
}

/**
 * Per-metric ordered source preference (highest first). The first source that
 * has a value for a given day wins.
 */
@Serializable
data class MetricSourcePriority(
    /** [MergeableMetric.raw] → ordered list of sources. */
    val order: Map<String, List<MetricSource>>,
) {
    fun sources(metric: MergeableMetric): List<MetricSource> =
        order[metric.raw] ?: metric.candidateSources

    /** Promote `source` to the top of `metric`'s ranking, keeping the rest in order. */
    fun settingPrimary(source: MetricSource, metric: MergeableMetric): MetricSourcePriority {
        val list = sources(metric).toMutableList()
        list.remove(source)
        list.add(0, source)
        return copy(order = order + (metric.raw to list))
    }

    /** `metric.raw -> [source.raw]`, for syncing to the backend. */
    val wire: Map<String, List<String>>
        get() = order.mapValues { (_, sources) -> sources.map { it.raw } }

    companion object {
        /** Defaults to each metric's candidate order, i.e. intervals.icu wins. */
        val default: MetricSourcePriority
            get() = MetricSourcePriority(MergeableMetric.entries.associate { it.raw to it.candidateSources })

        /** Rebuild from the wire form, filling any missing metric with defaults. */
        fun fromWire(wire: Map<String, List<String>>): MetricSourcePriority {
            val o = mutableMapOf<String, List<MetricSource>>()
            for ((metric, sources) in wire) {
                o[metric] = sources.mapNotNull { MetricSource.fromRaw(it) }
            }
            for (m in MergeableMetric.entries) {
                if (o[m.raw] == null) o[m.raw] = m.candidateSources
            }
            return MetricSourcePriority(o)
        }
    }
}
